### PR TITLE
Get rid of test front-loading in the duration_based_chunks algo

### LINKED ISSUE #26: Getting empty groups causing pytest to fail
Something in the most recent set of changes has broken something.  Previously I have been doing --splits 20 and all of the groups have at least one test.  Now, at least the last two groups are empty, which causes pytest to exit with error 5 and github actions to kill all of the tests.  Not sure what the precise cause of this is at the moment.

### PR DESCRIPTION
**Pull request by sondrelg**
Should hopefully resolve #26

### CODE REVIEW

**Comment by alexforencich** [comment]
Nope, still fails.  And if I change the test durations in that test case to all 1 except the last one which gets 100, all of the tests end up in the first group and I get 6 empty groups.

**Comment by sondrelg** [comment]
How would you suggest changing the implementation @alexforencich? I'm not seeing the issue 🙂

**Comment by alexforencich** [comment]
So I did not run the test correctly (have to actually install the package before you can run pytest).  After fixing that problem, the test case in the PR passes.  However, it fails if I change it to:

        durations = {}
        durations["a"] = 1
        durations["b"] = 1
        durations["c"] = 1
        durations["d"] = 1
        durations["e"] = 1
        durations["f"] = 1
        durations["g"] = 1
        durations["h"] = 100

In this case, only the first two groups are utilized and the rest are empty.

**Comment by alexforencich** [comment]
It seems to me like the goal of this algorithm is to preserve the test ordering.  So perhaps a better method would be to compute the cumulative sum into a new array, and then partition that on `total_duration*index/slices`.  Then adjust the indicies so they each correspond to at least one test.  

Incidentally, I think you can get slightly better performance out of least_duration by sorting the tests by duration first so they are processed from longest to shortest.  

However, the best option might be to do something like this: assign each test an ID number based on the original order, then sort based on duration, use the algorithm in least_duration to group the tests, then sort each group based on the ID.  This will preserve the test order, but produce more optimal grouping, and there will be no issue with empty groups unless there are more groups than tests (and in that case, what you can do is enable some random test in all of the otherwise-empty groups).

**Comment by sondrelg** [comment]
> So I did not run the test correctly (have to actually install the package before you can run pytest). After fixing that problem, the test case in the PR passes. However, it fails if I change it to:
> 
> ```
>     durations = {}
>     durations["a"] = 1
>     durations["b"] = 1
>     durations["c"] = 1
>     durations["d"] = 1
>     durations["e"] = 1
>     durations["f"] = 1
>     durations["g"] = 1
>     durations["h"] = 100
> ```
> 
> In this case, only the first two groups are utilized and the rest are empty.

That almost seems like a feature, not a bug; but more importantly, is that something that was changed in #14? We discussed fixing the change in behavior *and* the possibility of guaranteeing that every group has a test in #26, but to be clear - this PR just tries to fix the first point.

Would the test above behave differently in previous versions?

**Comment by alexforencich** [comment]
Well, if it wasn't causing github actions to crash, I wouldn't care as much.  But it is, so it's a bug.  Looks like yes, the splitting algorithm was rewritten in #14.  Looks like the previous version maybe was maybe duplicating tests in this case (not explicitly, but due to a side-effect of how it was written) so there weren't any empty groups, but this wasn't noticed since it didn't cause pytest to exit prematurely.

**Comment by mbkroese** [comment]
>Incidentally, I think you can get slightly better performance out of least_duration by sorting the tests by duration first so they are processed from longest to shortest.

@alexforencich you're right, that would be better. I just stuck with the current solution because that preserved relative order, which I thought was important for some. I'm happy to add this at a later stage.

**Comment by sondrelg** [comment]
Out of curiosity, why do people care about ordering? I really *only* want my test groups to be as even as possible, because the longest group sets the total time of my pipeline.

**Comment by mbkroese** [comment]
Just imagine you work for a company that has a large suite of tests and the corresponding technical debt. The tests already have order dependency (by accident) but it's too big a problem to fix all at once.

**Comment by sondrelg** [comment]
Yep lack of test isolation is the only thing I can think of 🙂

**Comment by alexforencich** [comment]
Well, you still might have ordering dependency issues as you're not going to run all of the tests.  

For relative order, you can restore that by assigning monotonically increasing IDs before doing the partitioning, then sort each group by this ID before running the tests.

**Comment by sondrelg** [comment]
Updated the PR now @alexforencich @jerry-git 

The `duration_based_chunks` algo should now have the following properties:

- Guarantees that every group is assigned a test (as long as `test count` > `splits`)
- Maintains order

**Comment by alexforencich** [comment]
Definitely looks a lot better. Still greedy I think, but not sure if there is a nice way of improving that without making it a lot more complicated.  At any rate, if you want the most even split, least_duration is probably a better choice.

Next step is to add some "safety" code to go ensure that there is at least one test enabled in each group, even if there are fewer tests than groups, by enabling a test in empty groups.  But that's not for this PR.

**Comment by sondrelg** [comment]
@jerry-git this is ready to review if you get a chance 🙂

**Comment by jerry-git** [comment]
Ahoy! Sorry this took a while, I was on vacation mode 🍹 I'll have a look.

File: tests/test_algorithms.py [hunk 1]
```diff
@@ -83,3 +83,51 @@ def test__split_tests_calculates_avg_test_duration_only_on_present_tests(self, a
         expected_first, expected_second = expected
         assert first.selected == expected_first
         assert second.selected == expected_second
+
+    def test_each_group_is_assigned_a_test_front_loaded(self):
+        from collections import namedtuple
+
+        from pytest_split import algorithms
```
Comment (jerry-git) [inline tests/test_algorithms.py:90]: Is there some motivation behind inline imports?

File: tests/test_algorithms.py [hunk 2]
```diff
@@ -83,3 +83,51 @@ def test__split_tests_calculates_avg_test_duration_only_on_present_tests(self, a
...
+        durations["a"] = 100
+        durations["b"] = 1
+        durations["c"] = 1
+        durations["d"] = 1
+        durations["e"] = 1
+        durations["f"] = 1
+        durations["g"] = 1
+        durations["h"] = 1
+
+        items = [item(x) for x in ["a", "b", "c", "d", "e", "f", "g", "h"]]
```
Comment (jerry-git) [inline tests/test_algorithms.py:104]: Would it be the same if the loop would be over `sorted(durations.keys())`?

File: tests/test_algorithms.py [hunk 3]
```diff
@@ -83,3 +83,51 @@ def test__split_tests_calculates_avg_test_duration_only_on_present_tests(self, a
...
+
+        groups = algorithms.duration_based_chunks(7, items, durations)
```
Comment (jerry-git) [inline tests/test_algorithms.py:106]: Why the first arg is 7 here and 8 in the next test?

File: tests/test_plugin.py [hunk 4]
```diff
@@ -61,20 +61,20 @@ class TestSplitToSuites:
...
         (3, 2, "least_duration", ["test_2", "test_5", "test_8"]),
         (3, 3, "least_duration", ["test_3", "test_6", "test_9"]),
-        (4, 1, "duration_based_chunks", ["test_1", "test_2", "test_3", "test_4"]),
-        (4, 2, "duration_based_chunks", ["test_5", "test_6", "test_7"]),
-        (4, 3, "duration_based_chunks", ["test_8", "test_9"]),
-        (4, 4, "duration_based_chunks", ["test_10"]),
+        (4, 1, "duration_based_chunks", ["test_1", "test_2", "test_3"]),
+        (4, 2, "duration_based_chunks", ["test_4", "test_5"]),
+        (4, 3, "duration_based_chunks", ["test_6"]),
+        (4, 4, "duration_based_chunks", ["test_7", "test_8", "test_9", "test_10"]),
```
Comment (jerry-git) [inline tests/test_plugin.py:77]: Doesn't this mean that the expected durations for each group will be:
* 1: 3s (1+1+1)
* 2: 2s (1+1)
* 3: 2s (2)
* 4: 8s (2+2+2+2)

If so, it doesn't feel right 🤔

File: tests/test_algorithms.py (hunk 2 above)
Comment (sondrelg) [reply tests/test_algorithms.py:92]: What do you mean, sorry?

File: tests/test_plugin.py (hunk 4 above)
Comment (sondrelg) [reply tests/test_plugin.py:77]: 1-5 are 1 second tests, while 5-10 are 2 second tests, so I believe the timings are 

- 3 seconds
- 3 seconds 
- 2 seconds 
- 8 seconds

And yes this is weird, but it's also what you should expect if you want to *stop adding tests before going over the threshold* of 15/4=3,75 seconds, **and** keeping absolute order 🤷

File: tests/test_algorithms.py (hunk 1 above)
Comment (sondrelg) [reply tests/test_algorithms.py]: moved 👍

File: tests/test_algorithms.py (hunk 3 above)
Comment (sondrelg) [reply tests/test_algorithms.py]: typo; fixed 👍

File: tests/test_algorithms.py (hunk 2 above)
Comment (alexforencich) [reply tests/test_algorithms.py:92]: I think he's saying you can do `items = sorted(durations.keys())` instead of duplicating all of the items.

File: tests/test_plugin.py (hunk 4 above)
Comment (alexforencich) [reply tests/test_plugin.py:77]: It's a very simple greedy algorithm; it's pretty much unavoidable that you're going to get strange far-from-optimal results like this.  The least_durations algorithm is much better and IMO should be the default and this algorithm should simply be removed.  But this algorithm does have the benefit of maintaining the sequence of tests, which I suppose could potentially be beneficial in some cases if there are ordering dependencies between tests.

Comment (alexforencich) [reply tests/test_plugin.py:77]: Actually, here's something to consider potentially: exceed the threshold, but keep track of a running total overage and use that when determining if you're over the threshold.  There are probably a few other heuristics that you could apply here to try to improve things a little bit.  For instance, something like this: ignore overage if the slot is empty and unconditionally assign, if it isn't empty then check to see if the resulting overage is less than half of the time required for that test, if so pack it, otherwise bump it to the next group.

File: tests/test_algorithms.py (hunk 2 above)
Comment (jerry-git) [reply tests/test_algorithms.py:92]: yeah I'm basically asking if this would be the same:
```python
items = [item(x) for x in sorted(durations.keys())]
```

IMO it would be better considering readability as then it's explicit that there's one item for each thingy in `durations`

File: tests/test_algorithms.py [hunk 5]
```diff
@@ -83,3 +83,27 @@ def test__split_tests_calculates_avg_test_duration_only_on_present_tests(self, a
...
+    def test_each_group_is_assigned_a_test_front_loaded(self):
+        item = namedtuple("item", "nodeid")
+
+        durations = {"a": 100, "b": 1, "c": 1, "d": 1, "e": 1, "f": 1, "g": 1, "h": 1}
+
+        items = [item(x) for x in ["a", "b", "c", "d", "e", "f", "g", "h"]]
+
+        groups = duration_based_chunks(8, items, durations)
+
+        for i in range(7):
```
Comment (jerry-git) [inline tests/test_algorithms.py:96]: Why 8 above and range(7) here in the loop below? Could we somehow get rid of the magical numbers and use e.g. variables which names would help the reader to understand what's going on?

File: tests/test_algorithms.py [hunk 6]
```diff
@@ -83,3 +83,27 @@ def test__split_tests_calculates_avg_test_duration_only_on_present_tests(self, a
...
+        item = namedtuple("item", "nodeid")
+
+        durations = {"a": 1, "b": 1, "c": 1, "d": 1, "e": 1, "f": 1, "g": 1, "h": 100}
+
+        items = [item(x) for x in ["a", "b", "c", "d", "e", "f", "g", "h"]]
+
+        groups = duration_based_chunks(8, items, durations)
+
+        for i in range(7):
+            assert groups[i].selected != []
```
Comment (jerry-git) [inline tests/test_algorithms.py:109]: Is this same as `assert all(group.selected for group in groups)`?

File: tests/test_plugin.py (hunk 4 above)
Comment (jerry-git) [reply tests/test_plugin.py:77]: >  The least_durations algorithm is much better and IMO should be the default and this algorithm should simply be removed

I have a use case which requires this plugin to support maintaining the order of the tests, which is why we need to keep `duration_based_chunks` or something similar. However, I can consider changing the default but that'd probably bite some other users of this plugin as well so need to consider that carefully 🙂

Comment (alexforencich) [reply tests/test_plugin.py:77]: Like I said, I would vote for removing it, but if we can't remove it for legacy reasons then we might as well spend a few cycles making it a little bit better.  I believe the current least_durations algorithm also preserves the ordering of the tests, but doesn't keep them contiguous - i.e. it could put 1, 3, 5, 6 in one split and 2, 4, 7, 8 in another split, but it won't ever make a group of 1, 5, 3, 6.  In contrast, duration_based_chunks keeps them contiguous and could generate 1, 2, 3 and 4, 5, 6, 7, 8, or something along those lines.  But both can break if, say, test 4 requires test 3 to run beforehand, because either algorithm could place those tests in different groups.

File: tests/test_algorithms.py (hunk 5 above)
Comment (alexforencich) [reply tests/test_algorithms.py:96]: Maybe use `for g in groups:` instead?

File: tests/test_algorithms.py (hunk 6 above)
Comment (alexforencich) [reply tests/test_algorithms.py:109]: That looks like it would probably work; maybe just add a comment to indicate that we're making sure there aren't any empty groups.  Or maybe explicitly check that the size of each group is nonzero.

File: tests/test_plugin.py (hunk 4 above)
Comment (jerry-git) [reply tests/test_plugin.py:77]: > But both can break if, say, test 4 requires test 3 to run beforehand, because either algorithm could place those tests in different groups.

True true, however duration_based_chunks most likely breaks less often, especially if there is a number of such dependencies between the tests 🙂

File: tests/test_algorithms.py (hunk 6 above)
Comment (jerry-git) [reply tests/test_algorithms.py:109]: >  we're making sure there aren't any empty groups ... check that the size of each group is nonzero

IMO `all(group.selected for group in groups)` does both and is explicit enough so no comments or further explanations would be needed. Just my opinion ofc 🙂

File: tests/test_plugin.py (hunk 4 above)
Comment (alexforencich) [reply tests/test_plugin.py:77]: Right, makes sense.  So, we should spend a few cycles improving the quality of the splitting in duration_based_chunks.  I think keeping track of the total "overage" while computing the splits should do quite a bit better.  If you wanted to do better than that, I think you might need to revisit the overall approach and perhaps come up with a less greedy algorithm.

Comment (jerry-git) [reply tests/test_plugin.py:77]: What ever we do here, at least my personal use case won't see much difference as it's thousands of tests (which each last 0-10 seconds) split into 3-5 groups, which is why I don't really have strong personal motivation behind improving the "general case". Thus, I think we could even go with this one from @sondrelg for now 🙂

File: tests/test_algorithms.py (hunk 2 above)
Comment (sondrelg) [reply tests/test_algorithms.py:92]: I just copied in the example test from [here](https://github.com/jerry-git/pytest-split/issues/26#issuecomment-865649265) and made sure we had one example that was extremely front-loaded and one that was extremely back-loaded. I can make it less explicit if you prefer.

File: tests/test_plugin.py (hunk 4 above)
Comment (sondrelg) [reply tests/test_plugin.py:77]: Feel free to jump in and make additional changes here - I don't have any need for this algo and was really only trying to help you out wrt. what was communicated in #27, so think I'll leave this as is 🙂

Comment (alexforencich) [reply tests/test_plugin.py:77]: Sounds like a reasonable plan to me - I'm also going to prefer the least_durations algorithm.  If I have a few minutes, perhaps I'll take a look at some tweaks to try to even things out a bit, but that can be done as a separate PR.

Comment (sondrelg) [reply tests/test_plugin.py:77]: I'll close this for now then 👍

Comment (alexforencich) [reply tests/test_plugin.py:77]: Well, without these changes, is duration_based_chunks still producing empty groups?  Or did that get fixed in a separate PR?

Comment (alexforencich) [reply tests/test_plugin.py:77]: OK, well my point was that even if the splitting isn't as perfect as it could be, at least this fix doesn't result in empty groups, so it should be merged and then possibly improved, instead of being closed.

Comment (sondrelg) [reply tests/test_plugin.py:77]: Sure thing. I've reopened it - feel free to make any additional changes you wish (cc @jerry-git) and merge. I'm on holiday now myself so likely won't be able to apply any code changes for a while.

**Comment by jerry-git** [comment]
Closing as stale, feel free to reopen 🙂
